File: leetcode/python/P0735_AsteroidCollision.py
# https://leetcode.com/problems/asteroid-collision/

from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        # when two asteroids of different signs are next to each other,
        # one or both explode.
        # we can put them in a queue (to keep order, i.e. vs stack)
        # when the signs differ --> explode (pop from queue, add remaining)

        # pay attention to:
        # - element in queue < 0; crt_elem > 0 --> doesn't collide
        # - when popping from queue, need variable to store the remaining asteroid
        # - stack (list) works as well; we don't need deque

        q = []

        for a in asteroids:
            if len(q) == 0:
                q.append(a)
            else:
                if len(q) > 0 and q[-1] * a > 0:
                    q.append(a)
                else:
                    crt = a
                    while len(q) > 0 and q[-1] > 0 and crt < 0:
                        prev_ = q.pop()
                        if abs(prev_) > abs(crt):
                            crt = prev_
                        elif abs(prev_) == abs(crt):
                            crt = 0

                    if crt != 0:
                        q.append(crt)

        return q

File: leetcode/cpp/P0735_AsteroidCollision.cpp
// https://leetcode.com/problems/asteroid-collision/

#include <algorithm>
#include <cmath>
#include <stack>
#include <vector>

class Solution {
 public:
  std::vector<int> asteroidCollision(std::vector<int>& asteroids) {
    std::stack<int> st;

    for (int x : asteroids) {
      int crt = x;
      bool both_explode = false;
      while (!st.empty()) {
        int top = st.top();
        both_explode = false;
        if (top > 0 && crt == -top) {
          both_explode = true;
          st.pop();
          break;
        } else {
          if (top < 0 || crt > 0) {
            break;
          }
        }

        // collision
        if (crt < 0 && top > 0) {
          st.pop();
          if (std::abs(crt) < std::abs(top)) {
            crt = top;
          }
        }
      }
      if (!both_explode) {
        st.push(crt);
      }
    }

    std::vector<int> result;
    while (!st.empty()) {
      int crt = st.top();
      result.push_back(crt);
      st.pop();
    }
    std::reverse(result.begin(), result.end());
    return result;
  }
};